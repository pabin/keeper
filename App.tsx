import React, { useState, useEffect } from 'react';
import { useColorScheme } from 'react-native';
import Toast from 'react-native-toast-message';

import { getObjectData } from './src/model/storageUtils';
import {
  createOrUpdateNoteDetails,
  updateNoteStatus,
} from './src/model/noteModel';
import StackNavigator from './src/navigation/StackNavigation';
import { themes } from './src/styles/colors';
import { Note } from './src/types/note';
import { NoteContextInterface } from './src/types/noteContext';

export const NoteContext = React.createContext<NoteContextInterface>([]);

const App = (): JSX.Element => {
  const [loading, setLoading] = useState<boolean>(true);
  const [notes, setNotes] = useState<Note[]>([]);
  const [favouriteNotes, setFavouriteNotes] = useState<Note[]>([]);
  const [archivedNotes, setArchivedNotes] = useState<Note[]>([]);

  // To Do: Theme switch
  const isDarkMode = useColorScheme() === 'dark';

  const getDatabase = async () => {
    const notesData: Note[] = await getObjectData('notes');

    if (notesData) {
      const favourites = notesData.filter(n => n.isFavourite);
      const archived = notesData.filter(n => n.isArchived);
      setNotes(notesData);
      setArchivedNotes(archived);
      setFavouriteNotes(favourites);
      setLoading(false);
    }
  };

  useEffect(() => {
    getDatabase();
  }, []);

  const onNoteStatusUpdate = async (note: Note, action: string) => {
    await updateNoteStatus(note, action);
    await getDatabase();
  };

  const onCreateOrUpdateNote = async (
    note: Note,
    title: string,
    body: string,
  ) => {
    await createOrUpdateNoteDetails(note, title, body);
    await getDatabase();
  };

  return (
    <>
      <NoteContext.Provider
        value={{
          notes,
          loading,
          favouriteNotes,
          archivedNotes,
          onNoteStatusUpdate,
          onCreateOrUpdateNote,
          theme: isDarkMode ? themes.dark : themes.light,
        }}
      >
        <StackNavigator />
        <Toast />
      </NoteContext.Provider>
    </>
  );
};

export default App;
