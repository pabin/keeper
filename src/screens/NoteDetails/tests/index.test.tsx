import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react-native';

import NoteDetail from '../index';
import {
  note,
  favouriteNote,
  archivedNote,
  notes,
} from '../../../fixtures/notes';
import { NoteContext } from '../../../../App';

describe('NoteDetail', () => {
  afterEach(cleanup);
  const onNoteStatusUpdate = jest.fn();

  const renderComponent = note => {
    const route = {
      params: { note },
    };
    return render(
      <NoteContext.Provider value={{ notes, onNoteStatusUpdate }}>
        <NoteDetail route={route} />
      </NoteContext.Provider>,
    );
  };

  test('renders correctly', () => {
    const { toJSON } = renderComponent(note);

    expect(toJSON()).toMatchSnapshot();
  });

  describe('marks favourite correctly', () => {
    test('when note is already favourite', () => {
      const { getByTestId } = renderComponent(favouriteNote);

      const markFavourite = getByTestId('markFavourite');
      fireEvent.press(markFavourite);
      expect(onNoteStatusUpdate).toHaveBeenCalledWith(
        favouriteNote,
        'favourite',
      );
    });

    test('when note is not favourite', () => {
      const notFavouriteNote = { ...favouriteNote, isFavourite: false };
      const { getByTestId } = renderComponent(notFavouriteNote);

      const markFavourite = getByTestId('markFavourite');
      fireEvent.press(markFavourite);
      expect(onNoteStatusUpdate).toHaveBeenCalledWith(
        notFavouriteNote,
        'favourite',
      );
    });
  });

  describe('marks archived correctly', () => {
    test('when note is already archived', () => {
      const { getByTestId } = renderComponent(archivedNote);

      const markArchived = getByTestId('markArchived');
      fireEvent.press(markArchived);
      expect(onNoteStatusUpdate).toHaveBeenCalledWith(note, 'archive');
    });

    test('when note is not archived', () => {
      const notArchivedNote = { ...archivedNote, isArchived: false };
      const { getByTestId } = renderComponent(notArchivedNote);

      const markArchived = getByTestId('markArchived');
      fireEvent.press(markArchived);
      expect(onNoteStatusUpdate).toHaveBeenCalledWith(
        notArchivedNote,
        'archive',
      );
    });
  });
});
