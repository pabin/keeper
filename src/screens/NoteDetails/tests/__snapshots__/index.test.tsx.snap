// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteDetail renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 16,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#263238",
        "borderRadius": 5,
        "elevation": 5,
        "margin": 1,
        "padding": 12,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      note title
    </Text>
    <Text
      style={
        Object {
          "fontSize": 10,
        }
      }
    >
      Fri Jan 21 2022 | 11:34:45 AM
    </Text>
  </View>
  <RCTScrollView
    removeClippedSubviews={true}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#263238",
        "borderRadius": 5,
        "elevation": 5,
        "marginVertical": 8,
        "minHeight": 350,
        "padding": 12,
      }
    }
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "paddingVertical": 12,
      }
    }
  >
    <View
      accessibilityLabel="Learn more"
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="markFavourite"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#01579B",
              },
            ]
          }
        >
          Mark as Favourite
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 12,
      }
    }
  >
    <View
      accessibilityLabel="Learn more"
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="markArchived"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#607D8B",
              },
            ]
          }
        >
          Mark as Archived
        </Text>
      </View>
    </View>
  </View>
</View>
`;
